iothub/device/src/Pipeline/ClientTransportStateMachine.cs

```diff
@@ -0,0 +1,96 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+
+namespace Microsoft.Azure.Devices.Client.Transport
+{
+    internal enum ClientTransportState
+    {
+        Closed,
+        Opening,
+        Open,
+        Closing,
+    }
+
+    internal enum ClientStateAction
+    {
+        OpenStart,
+        OpenSuccess,
+        OpenFailure,
+        CloseStart,
+        CloseComplete,
+        ConnectionLost,
+    }
+
+    internal class ClientTransportStateMachine
+    {
+        internal class StateTransition
+        {
+            private readonly ClientTransportState _currentState;
+            private readonly ClientStateAction _nextAction;
+
+            internal StateTransition(ClientTransportState state, ClientStateAction action)
+            {
+                _currentState = state;
+                _nextAction = action;
+            }
+
+            public override bool Equals(object obj)
+            {
+                return obj is StateTransition other
+                    && other._currentState == _currentState && other._nextAction == _nextAction;
+            }
+
+            public override int GetHashCode()
+            {
+                return _currentState.GetHashCode() + _nextAction.GetHashCode();
```

Comment:
https://learn.microsoft.com/en-us/dotnet/api/system.object.gethashcode?view=net-8.0#examples

```suggestion
                return _currentState.GetHashCode() ^ _nextAction.GetHashCode();
```

Reply:
I'm planning to use the prime number approach to evaluate the hash. That seems to be the most commonly recommended approach, and has been recommended by Jon Skeet as well: https://stackoverflow.com/questions/263400/what-is-the-best-algorithm-for-overriding-gethashcode/263416#263416.

It is easy to understand, doesn't have memory overhead and is fairly good with collisions.

The bit operations example in MS docs requires some shifting and wrapping of bits, so it'll end up requiring more lines of code that multiplication by prime numbers.

What do you think?

Reply:
Your hash set is very small. You likely don't have to worry about collisions. I'd pick a fast and cheap option. Bitwise operations are even faster than addition or multiplication.